<div style="border:solid green 2px; padding: 20px"> <h1 style="color:green; margin-bottom:20px">Reviewer's comment v1</h1>

Hello Spencer, my name is Dmitrii. I'm going to review your project! Nice to meet you! 🙌

You can find my comments under the heading **«Review»**. I will categorize my comments in green, blue or red boxes like this:

<div class="alert alert-success">
    <b>Success:</b> if everything is done successfully
</div>
<div class="alert alert-warning">
    <b>Remarks:</b> if I can give some recommendations or ways to improve the project
   
</div>
<div class="alert alert-danger">
    <b>Needs fixing:</b> if the block requires some corrections. Work can't be accepted with the red comments
</div>

Please don't remove my comments :) If you have any questions don't hesitate to respond to my comments in a different section. 
<div class="alert alert-info"> <b>Student comments:</b> For example like this</div>   
   

<div style="border:solid green 2px; padding: 20px">
<b>Reviewer's comment v1:</b>
    
<b>Overall Feedback</b> 
    

You've done a really good job overall! Your work shows that you understand the topic well and you've put in a lot of effort. There are a few small things to work on, but nothing too big:
    
    - Update the file path. 
    - Upate logic in the model evaluation stage.
    
Keep going like this, and with a little bit of tweaking, your work will be even better. I will wait for you to send me a new version of the project :)
    
</div>

<div style="border:solid green 2px; padding: 20px">
<b>Reviewer's comment v2:</b>
    
<b>Overall Feedback</b> 
    
Thank you for your comments and for going the extra mile in incorporating additional changes and improvements. 


Wish you cool projects in the next sprints! ☘️
    

</div>

# Project

# This is a project to develop a model that would analyze subscribers' behavior and recommend one of Megaline's newer plans: Smart or Ultra. 

# I have access to behavior data about subscribers who have already switched to the new plans. For this classification task, I will develop a model that will pick the right plan.   

# I will develop a model with the highest possible accuracy. In this project, the threshold for accuracy is 0.75. 

#  The data preprocessing step has already been completed, I will move straight to creating the model.  

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
It is always helpful for the reader to have additional information about project tasks. It gives an overview of what you are going to achieve in this project.


## importing packages I need

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

#### importing users_behavior dataset as df

In [2]:
df = pd.read_csv('users_behavior.csv')

<div class="alert alert-block alert-danger">
<b>Reviewer's comment v1</b>
 
Unfortunately, the file path is not correct here. Could you please check that? 

<div class="alert alert-info">  I uploaded the dataset to jupyter and that is how I called it. That is how I have been doing it throughout this course. I am not sure what is going wrong. When I hit Kernel and restart and run all it works. </div>   

<div class="alert alert-block alert-warning">
<b>Reviewer's comment v2</b>
 
This is a correct remote file path `'/datasets/users_behavior.csv'` 
    
You can use try/except structure in the following way to combine both paths: 
    
    
```
try:
    orders = pd.read_csv('users_behavior.csv', sep=';')

except FileNotFoundError:
    orders = pd.read_csv('/datasets/users_behavior.csv', sep=';')
```

<div class="alert alert-block alert-warning">
<b>Reviewer's comment v1</b>
 
It's a good practice to use `try/except blocks` when performing file operations or other tasks that might fail due to external reasons, such as the file not being present, issues with file permissions, or incorrect file formats. This way, you can handle errors gracefully and provide a more user-friendly error message, rather than having the program crash unexpectedly.

Here's how you can implement it:

```
try:
    orders = pd.read_csv(local_path['Churn.csv'], sep=';')

except FileNotFoundError:
    orders = pd.read_csv(server_path['Churn.csv'], sep=';')
```

<div class="alert alert-info">  I tried doing this but it said that the local_path wasn't defined or something like that </div>   

In [3]:
df.shape

(3214, 5)

#### there are 3214 rows and 5 columns

In [4]:
df.sample(12)

calls  minutes  messages   mb_used  is_ultra
1818   26.0   201.35      32.0   9027.27         1
860    72.0   410.23      68.0  16006.55         0
207    39.0   257.93      15.0   6572.61         0
3075   39.0   278.79      21.0  11430.52         0
923    61.0   421.74      41.0  19746.94         0
1340   32.0   244.80      13.0   7213.58         0
1305   41.0   302.04     106.0   9755.95         0
227   111.0   717.16     191.0  38287.82         1
2978   45.0   270.14      29.0      0.00         1
2993   45.0   358.45      55.0  12715.68         0
651    18.0   147.93      18.0   1904.01         1
1215   53.0   354.85      10.0  21820.27         0

<div class="alert alert-warning">
<b>Reviewer's comment v1:</b>
    
Before using data for any ML model, it could be helpful to check for duplicates, verify the uniqueness of values, and identify any nulls.
    
Consider visualizing some key aspects of the data (e.g., distribution of variables, correlation matrix) to provide a better understanding of the dataset.

<div class="alert alert-info">  I did not do any of this because the project description says "Since you’ve already performed the data preprocessing step, you can move straight to creating the model."  </div>  

<div class="alert alert-success">
<b>Reviewer's comment v2:</b>
    
👍

#### this is a classification model. is_ultra will be my target. calls, minutes, messages, mb_used will be my features

In [5]:
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra']

In [6]:
features.shape

(3214, 4)

In [7]:
target.shape

(3214,)

#### features df has 4 columns meaning the is_ultra column has been dropped and the target df only has 1 column, is_ultra

#### splitting the data into training and validation sets

In [8]:
df_train, df_valid = train_test_split(df, test_size=0.25, random_state=12345, stratify=df['is_ultra'])
features_train = df_train.drop(['is_ultra'], axis=1)
target_train = df_train['is_ultra']
features_valid = df_valid.drop(['is_ultra'], axis=1)
target_valid = df_valid['is_ultra']

In [9]:
print(features_train.shape)
print(target_train.shape)
print(features_valid.shape)
print(target_valid.shape)

(2410, 4)
(2410,)
(804, 4)
(804,)


<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
The data split is correct.  However, consider using `stratify` parameter in train_test_split to maintain the distribution of the target variable, especially important for imbalanced datasets.

#### testing the decision tree classifier model with different depths for highest accuracy

In [10]:
for depth in range(1, 15):
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345) 
    model.fit(features_train, target_train)  
    predictions_valid = model.predict(features_valid)
    print("max_depth =", depth, ": ", end='')
    print(accuracy_score(target_valid, predictions_valid))

max_depth = 1 : 0.7325870646766169
max_depth = 2 : 0.7674129353233831
max_depth = 3 : 0.7723880597014925
max_depth = 4 : 0.7960199004975125
max_depth = 5 : 0.7860696517412935
max_depth = 6 : 0.7922885572139303
max_depth = 7 : 0.7960199004975125
max_depth = 8 : 0.7898009950248757
max_depth = 9 : 0.7960199004975125
max_depth = 10 : 0.7960199004975125
max_depth = 11 : 0.7910447761194029
max_depth = 12 : 0.7835820895522388
max_depth = 13 : 0.7748756218905473
max_depth = 14 : 0.7798507462686567


#### max depth 7 got the highest accuracy with .7898

#### even though this meets the threshold, there might be a better model with a higher accuracy score

#### testing the accuracy of the Random Forest Classifier model

In [11]:
best_score = 0
best_est = 0
for est in range(1, 15):
    model = RandomForestClassifier(random_state=12345, n_estimators=est) 
    model.fit(features_train, target_train) 
    score = model.score(features_valid, target_valid) 
    if score > best_score:
        best_score = score
        best_est = est

print("Accuracy of the best model on the validation set (n_estimators = {}): {}".format(best_est, best_score))


Accuracy of the best model on the validation set (n_estimators = 10): 0.7985074626865671


#### Random Forest model has a higher accuracy than the Decision Tree model

#### Testing Logistic Regression Model for highest acccuracy

In [12]:
model =  LogisticRegression(random_state=12345, solver='liblinear')
model.fit(features_train, target_train)  
score_train = model.score(
    features_train, target_train
)  
score_valid = model.score(features_valid, target_valid)


print(
    "Accuracy of the logistic regression model on the training set:",
    score_train,
)
print(
    "Accuracy of the logistic regression model on the validation set:",
    score_valid,
)

Accuracy of the logistic regression model on the training set: 0.7153526970954357
Accuracy of the logistic regression model on the validation set: 0.7114427860696517


<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Everything is correct here! Great that you've managed to check multiple models. 



#### Logistic Regression has the lowest accuracy score

#### I will be using the Random Forest Model

In [13]:
final_model = RandomForestClassifier(random_state=12345, n_estimators=12)

#### checking the quality of the model using the test set

In [14]:
final_model.fit(features_train, target_train)

RandomForestClassifier(n_estimators=12, random_state=12345)

In [15]:
test_predictions = final_model.predict(features_valid)

def accuracy(answers, predictions):
    correct = 0
    for i in range(len(answers)):
        if answers.iloc[i] == predictions[i]:  
            correct += 1
    return correct / len(answers)

print('Accuracy:', accuracy(target_valid, test_predictions))

Accuracy: 0.7935323383084577


#### after training the model on the training set and testing it on the valid set I received an accuracy score of 79%

<div class="alert alert-danger">
<b>Reviewer's comment v1:</b>
    
You've trained your model on the entire dataset `(features, target)` and then evaluated it on the same dataset. This practice leads to overly optimistic performance estimates because the model is being tested on data it has already seen. The high accuracy score of nearly 100% is thus not a reliable indicator of how well your model would perform on truly unseen data.

You should fit your model using only the `features_train` and `target_train` datasets.
After fitting the model, you make predictions on the `features_valid` dataset, which the model has not seen during training.
The accuracy (or any other performance metric) should then be calculated based on how well the model's predictions on the validation set (`features_valid`) match the actual targets in target_valid.

<div class="alert alert-success">
<b>Reviewer's comment v2:</b>

Everything is correct now. 

<div class="alert alert-warning">
<b>Reviewer's comment v1:</b>
    
In our case, we have a very skewed distribution of users across the two plans. As you can see, only 30% are enrolled in the Ultra plan, therefore it could be usefull to have a dummy model in place to validate results of the RandomForestClassifier.  
    
You could use `DummyClassifier` to implement this sanity check. You could read about it [here](https://scikit-learn.org/stable/modules/generated/sklearn.dummy.DummyClassifier.html)
    
```
from sklearn.dummy import DummyClassifier
# Initialize the DummyClassifier to predict the most frequent class
dummy_clf = DummyClassifier(strategy="most_frequent", random_state=0)
# Fit the dummy classifier on the training data
...
```

# In conclusion, the model Random Forest Classifier  with the number of estimators set to 12 will give almost 80% accuracy in determining what plan to offer a customer based off of what their data shows on their calls, minutes, messages, and mb used. 

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    

Great job on your overall conclusions and recommendations!  Your recommendations are well-thought and could be very valuable to the business.